[Add Feature] Simulate Pulse-Spillover noise in the custom backend for pulse level simulation

lukasvoss opened the pull request:
This PR introduces the option of passing a `Dict[Tuple[int, int], float] ` to the `custom_backend` function that stores noise-couplings strengths between qubits. We are modeling pulse-spillover noise.

If not specified otherwise, the noise profile is considered to be **symmetrical** allowing to only pass one noise-direction $(i, j): \gamma$ without the need to also pass the counter-direction $(j, i): \gamma$.

arthurostrauss commented on template_configurations/qiskit/pulse_level/dynamics_backends/noisy_circuit/noisy_dynamics_config.py:
```diff
@@ -0,0 +1,174 @@
+
+    dt = 2.2222e-10
+
+    jax_solver = JaxSolver(
```
Same argument here for what follows, since this is a replication of the previous code, should be embedded into a function to be called again. The goal is to avoid copy past as much as possible.

arthurostrauss commented on helper_functions.py:
Those added functions are not really necessary for the whole functioning of the workflow. They should therefore be packed in another utils.py present in the dynamics_backend folder in template configurations instead.

arthurostrauss commented on helper_functions.py:
```diff
@@ -1508,7 +1508,167 @@ def cost_function(x):
+    neighbor relationships among them. Each key in the noise_couplings dictionary is a tuple
+    representing a pair of qubits, and its corresponding value indicates the coupling strength
+    between these qubits. Only positive coupling strengths are considered as valid connections.
+    Self-loops can optionally be included if they exist and have positive coupling strength.
```
What would self loop correspond to? Some distortion on the baseline pulse? If yes, this will not be very be useful if it's not stochastic. We could however think about how to implement another noise model which implements the signal with some white noise. Food for thought for later.

lukasvoss replied:
Agreed, I removed this. Let's dedicate another model to imperfect pulses at some point in the future.

arthurostrauss commented on helper_functions.py:
```diff
@@ -1508,7 +1508,167 @@ def cost_function(x):
+    This function processes a dictionary of noise couplings between qubits to determine the
+    neighbor relationships among them. Each key in the noise_couplings dictionary is a tuple
+    representing a pair of qubits, and its corresponding value indicates the coupling strength
+    between these qubits. Only positive coupling strengths are considered as valid connections.
```
You should explain how those couplings differ from the original ones present in the baseline function. This is a coupling that exists only when a pulse is played no?

lukasvoss replied:
Added some clarifying documentation.

arthurostrauss commented on helper_functions.py:
```diff
@@ -1773,27 +1943,36 @@ def create_circuit_from_own_unitaries(circuit_context: QuantumCircuit, **kwargs)
     """
     new_circuit = QuantumCircuit(circuit_context.num_qubits)
     gate_count = {}
-    target_gate_instruction = kwargs.get('gate')
```
The function create_circuit_from_own_unitaries should also be part of another dedicated utils file for the gate level. Not necessary for the functioning of the workflow.

arthurostrauss commented on helper_functions.py:
```diff
@@ -1773,27 +1943,36 @@ def create_circuit_from_own_unitaries(circuit_context: QuantumCircuit, **kwargs)
 
-def process_gate(inst: Instruction, qargs: Qubit, gate_count: Dict, circuit_context: QuantumCircuit):
+
+def process_gate(
```
Same comment for going into another utils file for this method

arthurostrauss commented on template_configurations/qiskit/pulse_level/dynamics_backends/dynamics_config.py:
```diff
@@ -6,6 +6,13 @@
 from qiskit_dynamics import DynamicsBackend, Solver
 from custom_jax_sim import JaxSolver
 import numpy as np
+from helper_functions import (
```
Okay but should be imported from dedicated utils.

arthurostrauss commented on template_configurations/qiskit/pulse_level/dynamics_backends/noisy_circuit/noisy_dynamics_config.py:
```diff
@@ -6,7 +6,14 @@
 from custom_jax_sim import JaxSolver
 import numpy as np
-from helper_functions import get_noise_coupling_neighbours
+from helper_functions import (
```
Same comment as before (import from utils)

arthurostrauss commented on template_configurations/qiskit/pulse_level/dynamics_backends/noisy_circuit/noisy_dynamics_config.py:
```diff
@@ -27,7 +32,10 @@ def custom_backend(
-    Custom backend for the dynamics simulation.
+    Custom noisy backend for the dynamics simulation.
+    We allow for ZZ-crosstalk noise couplings between neighbouring qubits modeling hardware noise in transmon architectures.
+    This noise occurs when pulses are applied to one qubit and affect the neighbouring qubits as well due to wires being close to each other.
```
I'm confused. This noise is the pulse spillover right? The ZZ crosstalk interaction is part of the static hamiltonian, and not part of the driving operators.

lukasvoss replied:
Yes, it wrote the wrong noise type here, it should say pulse-spillover. Updated the docstring.

arthurostrauss reviewed (changes requested):
Thanks for the quick edits. Just have to put all the added functions to another dedicated utils file, and also precise that this is pulse spillover and not static ZZ crosstalk (unless I'm mistaken). 
You can do the same reasoning for the functions that are used for custom Aer backends with gate level custom noise models.

lukasvoss commented:
First things first: The features introduced are for modeling pulse-spillover noise arising from control lines in real hardware being very close to each other that neighbouring qubits can be effected when playing a pulse on a certain qubit.
Passing the dictionary of couplings, we can specify between which qubit pairs this noise arises and by what spillover rate $\gamma$ the qubit is affected.

I moved the responsible functions into a utils file and updated documentation.

arthurostrauss commented on template_configurations/qiskit/gate_level/noisy_circuit/utils.py:
```diff
@@ -0,0 +1,93 @@
+
+        # Check if the current gate matches the target operation
+        if (
+            isinstance(target_gate_instruction, Instruction)
```
Could the elements inside the circuit.data be something else than Instruction ? Can't think of an example now but I'd be happy to hear about it.

arthurostrauss commented on template_configurations/qiskit/pulse_level/dynamics_backends/utils.py:
```diff
@@ -0,0 +1,191 @@
+    """
+    This function describes the noise effect of pulse-spillover for qubit-pairs specified in noise_couplings. This noise only occurs when a pulse (intended for qubit i) leaks
+    into another qubit j. So, it only affects the drive operators of a qubit.
+    This is a common noise source in superconducting qubits.
```
Given the poor literature I could find on this noise source I would not say it is considered common. The question of it being neglected now because it is solved or because it is just not well studied is not answered yet?

arthurostrauss reviewed (changes requested):
Just an adjustment in the docstrings and we should be good to go.

Good work!
